File: leetcode_2/jh.go
package leetcode_2

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

// https://leetcode.com/problems/add-two-numbers/
// Runtime 3ms Beats 98.8% Memory 4.6 MB Beats 87.88%
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	overFlowFlag := false
	newListNode := new(ListNode)
	current := newListNode
	for {
		sum := getNodeVal(l1) + getNodeVal(l2)
		if overFlowFlag {
			sum += 1
			overFlowFlag = false
		}

		if sum >= 10 {
			sum = sum % 10
			overFlowFlag = true
		}
		current.Val = sum

		l1 = getNext(l1)
		l2 = getNext(l2)

		if isNil(l1, l2) && !overFlowFlag {
			break
		}

		current.Next = new(ListNode)
		current = current.Next
	}

	return newListNode
}

func getNext(node *ListNode) *ListNode {
	if node == nil {
		return nil
	}
	return node.Next
}

func getNodeVal(node *ListNode) int {
	if node == nil {
		return 0
	}

	return node.Val
}

func addTwoNumbersVersion2(l1 *ListNode, l2 *ListNode) *ListNode {
	overFlowFlag := false
	newListNode := new(ListNode)
	current := newListNode
	for {
		sum := 0
		if isNotNil(l1) {
			sum += l1.Val
			l1 = l1.Next
		}

		if isNotNil(l2) {
			sum += l2.Val
			l2 = l2.Next
		}

		if overFlowFlag {
			sum += 1
			overFlowFlag = false
		}

		if sum >= 10 {
			sum = sum % 10
			overFlowFlag = true
		}
		current.Val = sum

		if isNil(l1, l2) && !overFlowFlag {
			break
		}

		current.Next = new(ListNode)
		current = current.Next
	}

	return newListNode
}

func isNil(nodes ...*ListNode) bool {
	for _, node := range nodes {
		if node != nil {
			return false
		}
	}
	return true
}

func isNotNil(nodes ...*ListNode) bool {
	return !isNil(nodes...)
}

File: leetcode_21/jh.py
# https://leetcode.com/problems/merge-two-sorted-lists/description/
# Runtime 51 ms Beats 35.52% Memory 16.4 MB Beats 19.46%
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        new_list = ListNode()
        current = new_list
        while self.is_not_none(list1) and self.is_not_none(list2):
            if list1.val < list2.val:
                current.next = list1
                list1 = list1.next
            else:
                current.next = list2
                list2 = list2.next
            current = current.next
        
        if self.is_not_none(list1):
            current.next = list1
        if self.is_not_none(list2):
            current.next = list2

        return new_list.next


    def is_not_none(self, elem: Optional[ListNode]) -> bool :
        return elem is not None
    